internal/adapters/dataproviders/database/mongo/common.go

```diff
@@ -74,3 +78,200 @@ func (c *Connection) CheckConnection(ctx context.Context) bool {
+	ctx context.Context,
+	conn *Connection,
+	startDate, endDate time.Time,
+	collectionName string,
+	mapper func(bson.D) Q,
+) ([]Q, []string, error) {
+	quoteFilter := bson.D{
+		{Key: "agreement_timestamp", Value: bson.D{
+			{Key: "$gte", Value: startDate.Unix()},
+			{Key: "$lte", Value: endDate.Unix()},
+		}},
+	}
+	var storedQuotes []bson.D
+	quoteCursor, err := conn.Collection(collectionName).Find(ctx, quoteFilter)
+	if err != nil {
+		return nil, nil, err
+	}
+	if err = quoteCursor.All(ctx, &storedQuotes); err != nil {
+		return nil, nil, err
+	}
+	quoteHashes := make([]string, 0, len(storedQuotes))
+	quotes := make([]Q, 0, len(storedQuotes))
+	for _, stored := range storedQuotes {
+		quoteObj := mapper(stored)
+		quotes = append(quotes, quoteObj)
+		hashValue, ok := getStringValueFromBSON(stored, "hash")
+		if ok {
+			quoteHashes = append(quoteHashes, hashValue)
+		}
+	}
+	return quotes, quoteHashes, nil
+}
+
+func getStringValueFromBSON(doc bson.D, key string) (string, bool) {
+	data, err := bson.Marshal(doc)
+	if err != nil {
+		return "", false
+	}
+	rawValue := bson.Raw(data).Lookup(key)
+	return rawValue.StringValueOK()
+}
+
+type QuoteHashProvider interface {
+	GetQuoteHash() string
+}
+
+func fetchRetainedQuotes[R QuoteHashProvider](
+	ctx context.Context,
+	conn *Connection,
+	startDate, endDate time.Time,
+	collectionName string,
+	existingQuoteHashes []string,
+) ([]R, []string, error) {
+	retainedFilter := createRetainedFilter(startDate, endDate, existingQuoteHashes)
+	var retainedQuotes []R
+	retainedCursor, err := conn.Collection(collectionName).Find(ctx, retainedFilter)
+	if err != nil {
+		return nil, nil, err
+	}
+	if err = retainedCursor.All(ctx, &retainedQuotes); err != nil {
+		return nil, nil, err
+	}
+	additionalHashes := findAdditionalQuoteHashes(retainedQuotes, existingQuoteHashes)
+	return retainedQuotes, additionalHashes, nil
+}
+
+func createRetainedFilter(startDate, endDate time.Time, quoteHashes []string) bson.D {
+	return bson.D{
+		{Key: "$or", Value: bson.A{
+			bson.D{{Key: "quote_hash", Value: bson.D{
+				{Key: "$in", Value: quoteHashes},
+			}}},
+			bson.D{
+				{Key: "created_at", Value: bson.D{
+					{Key: "$gte", Value: startDate.Unix()},
+					{Key: "$lte", Value: endDate.Unix()},
+				}},
+			},
+		}},
+	}
+}
+
+func findAdditionalQuoteHashes[R QuoteHashProvider](retainedQuotes []R, existingQuoteHashes []string) []string {
+	existingMap := make(map[string]bool, len(existingQuoteHashes))
+	for _, hash := range existingQuoteHashes {
+		existingMap[hash] = true
+	}
+	additionalMap := make(map[string]bool)
+	for i := range retainedQuotes {
+		hash := retainedQuotes[i].GetQuoteHash()
+		if !existingMap[hash] {
+			additionalMap[hash] = true
+		}
+	}
+	additionalHashes := make([]string, 0, len(additionalMap))
+	for hash := range additionalMap {
+		additionalHashes = append(additionalHashes, hash)
+	}
+	return additionalHashes
+}
+
+func fetchAdditionalQuotes[Q any](
```

Comment:
my comment for this function is the same as for the other functions in this file